Feature/support enumeration type hints

lorenzocelli opened the pull request:
Issue #10

siliconlad reviewed (commented):
One thing I've realized is the following.

```python
from enum import Enum


class IntEnum(Enum):
    A = 1
    B = 2

IntEnum.A == 1  # False
```

Which means that when the LLM calls a function which accepts a type of `IntEnum`, it will fail because `IntEnum.A` is not the same as `1`, for example.

To me there are two possible options:
1. We could warn the user in the documentation that they need to deal with this. Not ideal but the fix would just be to add `IntEnum(1)`.
2. We could wrap the provided function to do the conversion ourselves. This would obviously be more convenient, but wonder if that introduces other problems.

What do you think @LorenzoCelli

LorenzoCelli commented:
> One thing I've realized is the following.
> 
> ```python
> from enum import Enum
> 
> 
> class IntEnum(Enum):
>     A = 1
>     B = 2
> 
> IntEnum.A == 1  # False
> ```
> 
> Which means that when the LLM calls a function which accepts a type of `IntEnum`, it will fail because `IntEnum.A` is not the same as `1`, for example.
> 
> To me there are two possible options:
> 
> 1. We could warn the user in the documentation that they need to deal with this. Not ideal but the fix would just be to add `IntEnum(1)`.
> 2. We could wrap the provided function to do the conversion ourselves. This would obviously be more convenient, but wonder if that introduces other problems.
> 
> What do you think @LorenzoCelli

Right. I agree with you, the first option is not ideal. 

I think we could implement the second more easily if we stored the list of `ParameterSchema` objects in `FunctionSchema`. Then we could do something like the following inside `_GPTEnabled.__call__` (ignoring args here):

```python
def __call__(self, *args, **kwargs):

    p_schemas: list[ParameterSchema] = self.schema.parameters

    for schema in p_schemas:
        if isinstance(schema, ListParameterSchema):
            # replace value with corresponding enum type
            kwargs[schema.parameter.name] = IntEnum(kwargs[schema.parameter.name])

    return self.func(*args, **kwargs)
```

We could also delegate to `ParameterSchema` the task of converting a value back and remove the call to `isinstance`.

siliconlad commented:
Yeah I think that sounds good!

LorenzoCelli commented:
We're still missing the value to enum conversion on invocation

siliconlad commented:
@LorenzoCelli I've marked this as a draft. Let me know when it's ready for me to review 👍

LorenzoCelli commented:
@siliconlad when we convert the enum to a set of values to send to the api, do you think we should keep the values or the names? Names are usually more informative than values, such as in this example from the python docs:

```python
class Color(Enum):
    RED = 1
    GREEN = 2
    BLUE = 3
```

LorenzoCelli commented:
In order to store the parameter schemas inside the function schema class, I found it convenient to remove the static methods in favour of a more instance-oriented approach. For example `FunctionSchema.schema(...) -> dict` becomes  `mySchema._populate_schema(...)` called in the constructor to populate the schema dictionary. In this way it is easier to store other information like `parameter_schemas` in the process.

Another way could be to keep the static methods and provide them with a `parameter_schemas` dictionary to fill.

siliconlad commented:
> @siliconlad when we convert the enum to a set of values to send to the api, do you think we should keep the values or the names? Names are usually more informative than values, such as in this example from the python docs:
> 
> ```python
> class Color(Enum):
>     RED = 1
>     GREEN = 2
>     BLUE = 3
> ```

The LLM might perform better with the names I think.

LorenzoCelli commented:
> The LLM might perform better with the names I think.

I suspected, I will mark this as a draft again then

siliconlad commented on tool2schema/schema.py:
```diff
@@ -83,6 +83,14 @@ def __init__(self, func, **kwargs) -> None:
         functools.update_wrapper(self, func)
 
     def __call__(self, *args, **kwargs):
+
+        for key in kwargs:
+            if key in self.schema.parameter_schemas:
+                # Convert the JSON value to the type expected by the method
+                kwargs[key] = self.schema.parameter_schemas[key].parse_value(
+                    kwargs[key]
+                )
+
```
There should be a way to also check `args` so you can also support the following test case:

```python
a, _, _, _ = function_custom_enum(CustomEnum.A.name, "", False, [])
assert a == CustomEnum.A
```

Unless you think this is something we shouldn't do?

LorenzoCelli replied:
I did it this way because I assumed that when the function is invoked with data coming from the API we only use keyword arguments. But if you think it's useful in some case to parse args as well I will do it

siliconlad replied:
I think it will be good for completeness.

LorenzoCelli replied:
Will do

siliconlad commented on tool2schema/schema.py:
```diff
@@ -105,108 +113,106 @@ def wrapper(function):
         """
         if schema_type == SchemaType.TUNE:
-            return FunctionSchema.schema(self.f, schema_type)["function"]
+            return FunctionSchema(self.f, schema_type).to_json()["function"]
```
This works, but it feels like a slight misdirection (or slightly abusing the `SchemaType` enum).

What do you think about populating `self.parameter_schemas` within the constructor but waiting to actually create the json in the `to_json` method? That way you don't need to create an entirely new object (redoing some computation) to create a slightly different JSON format.

LorenzoCelli replied:
I agree with you and we can improve it, I did it like this to mimic the existing code:
```python
    def to_json(self, schema_type: SchemaType = SchemaType.API) -> dict:
        if schema_type == SchemaType.TUNE:
            # we invoke the static method and re-compute the json
            return FunctionSchema.schema(self.f, schema_type)["function"]
        return self.schema
```

I will change it to compute the parameters on initialisation and the json dictionary only upon request. Maybe I can imitate the `_add_...` methods of ParameterSchema.

It would also be beneficial because currently we compute the schema dictionary for every function that has a decorator as soon as the decorator is instantiated, while in this way the schema would not be computed until the user invokes `to_json` (also much easier to debug).

siliconlad replied:
> I did it like this to mimic the existing code

Oops haha.

LorenzoCelli replied:
Checked in, let me know what you think :)

siliconlad replied:
Skimmed through it and it looks good! Will have a proper look through when it's ready.

siliconlad reviewed (changes requested):
A couple of comments :)

LorenzoCelli commented:
Refactored ParameterSchema implementation to be closer to FunctionSchema implementation (using `get_...()` methods)

LorenzoCelli commented:
To support `Enum` parameters with default values, I added a `ParameterSchema.encode_value` method to convert an instance to its json representation (in the case of `Enum`, the name). Contextually I renamed `parse_value` (which converts the json representation to an instance) to `decode_value`.

siliconlad commented on tests/test_schema.py:
```diff
@@ -161,29 +162,37 @@ def remove_param(self, param: str) -> None:
         """
         self.schema["function"]["parameters"]["properties"][param] = value
 
+    def get_required_parameters(self) -> list[str]:
```
Should use `Optional` here because technically `get` can return `None` I think.

```suggestion
    def get_required_parameters(self) -> Optional[list[str]]:
```

lorenzocelli replied:
Yep, done in e72566f

siliconlad commented on tool2schema/parameter_schema.py:
```diff
@@ -36,60 +40,96 @@ def matches(parameter: Parameter) -> bool:
+            "enum": self._get_enum,
+        }
+
+        json = dict()
+
+        for field in fields:
+            if (value := fields[field]()) != Parameter.empty:
+                json[field] = value
```
Could replcae with a [dictionary comprehension](https://stackoverflow.com/questions/22244074/iterate-over-a-dictionary-by-comprehension-and-get-a-dictionary). Which may or may not be clearer.

lorenzocelli replied:
Done in ad025df

siliconlad commented on tool2schema/parameter_schema.py:
```diff
@@ -36,60 +40,96 @@ def matches(parameter: Parameter) -> bool:
         raise NotImplementedError()
 
-    def _add_type(self, schema: dict):
+    def _get_type(self) -> Union[str, Parameter.empty]:
```
What is the benefit of using `Parameter.empty` vs just `Optiona[str]`?

lorenzocelli replied:
I used it to differentiate between "I return None" and "I return nothing, discard it" in `_get_default`, since the default value might be none and still be included in the schema. In the other methods it is just for uniformity, it simplifies the iteration in `to_json`

siliconlad replied:
Maybe add a comment about this somewhere?

lorenzocelli replied:
It says something in the docstring but I added a comment with 6d659b4 to make it clearer

siliconlad commented on tool2schema/parameter_schema.py:
```diff
@@ -145,21 +187,94 @@ def matches(parameter: Parameter) -> bool:
+
+        :param value: The enum instance to be converted.
+        """
+        return value.name
```
Might be worth checking the type before calling `value.name`?

lorenzocelli replied:
Done in ad025df, replaced with 
```python
return value.name if isinstance(value, Enum) else value
```
so that the user can write `def myFunc(b: CustomEnum = "B")` and the method does not throw an exception

siliconlad reviewed (changes requested):
Couple of small comments. Otherwise looks good 🚀

lorenzocelli commented:
I updated the 'enumerations' section of the README

siliconlad commented on tool2schema/parameter_schema.py:
```diff
@@ -36,60 +40,92 @@ def matches(parameter: Parameter) -> bool:
-        self._add_items(json)
-        self._add_type(json)
-        self._add_enum(json)
+        fields = {
+            "description": self._get_description,
+            "default": self._get_default,
+            "items": self._get_items,
+            "type": self._get_type,
+            "enum": self._get_enum,
+        }
+
+        json = {field: value for field in fields if (value := fields[field]()) != Parameter.empty}
```
I feel like this is simpler?

```suggestion
        fields = {
            "description": self._get_description(),
            "default": self._get_default(),
            "items": self._get_items(),
            "type": self._get_type(),
            "enum": self._get_enum(),
        }

        json = {f: v for f, v in fields.items() if v != Parameter.empty}
```

siliconlad reviewed (approved):
LGTM!
